ENH: Use Dragon4 algorithm to print floating values

ahaldane opened the pull request:
This is a C port of [Ryan Juckett's C++ implementation (link)](http://www.ryanjuckett.com/programming/printing-floating-point-numbers/part-4/) of the Dragon4 algorithm for printing floating-point values to numpy. See earlier discussion in #9919 and #9932, which are both replaced by this PR. Please check the license at the top of the dragon4.c file to confirm it is suitable.

This is a WIP and I am still deciding some of the behavior, but it is in "working" state. I am putting it up now to run tests, and if anyone wants to try it out. Here's a quick summary:

 * Adds a python interface to the Dragon4 algorithms in `np.core.multiarray.dragon4_positional` and `np.core.multiarray.dragon4_scientific`. Documentation of arguments is in the C code, but you can call it with a single python floating scalar argument to try it out.
 * Numpy scalars use the dragon4 algorithm in "unique" mode (see below) for str/repr, in a way that tries to match python float output. Note: one possible bugfix is that the scientific mode cutoff for all floats is now 1e16, instead of 1e10 for float32 and 1e17 for float64 like before. CPython's cutoff is 1e16. This also fixes complex reprs like in #9332.
 * updates `arrayprint.py` code to use dragon4. I also added a new option, that you can set `np.printoptions(precision='unique')`, which will print the minimum number of digits to uniquely specify the value. I left the default at `precision=8` for now. 
 * In `arrayprint.py` I've also stolen/reworked the `FormatFloat` generalizations from #9919. Now all floating types just use a single `FloatingFormat` formatter, which calls the dragon4 routines.

Some examples that you can compare to the old behavior:
```python
print(complex(0,np.inf), np.complex128(complex(0,np.inf)), str(np.array([np.complex128(complex(0,np.inf))])))
np.set_printoptions(precision='unique')
print(np.arange(10, dtype='f4')/10.)
a = np.random.rand(4)
print(repr(a))
print(repr(a.astype('f4')))
print(repr(a.astype('f2')))
print(repr(10.**np.arange(20)))
```

I still need to do some cleanups and tests. Also, there may be a bug in the original implementation because `np.float64(1e23)` isn't nicely rounded, unlike the other powers of 10.

ahaldane committed:
ENH: Port the Dragon4 implementation to C/Numpy

Port, plus modifications, of Ryan Juckett's implementation of the
Dragon4 algorithm for printing floating point values, from C++ to C.

ahaldane committed:
ENH: Add python interface to dragon4 algorithm

ahaldane committed:
ENH: Make numpy float scalars use dragon4 output

ahaldane committed:
ENH: Make arrayprint code use dragon4 algorithm

ahaldane committed:
TST: update tests to account for dragon4 fixes

ahaldane commented:
Output of the code in my last comment, with this PR:
```
(infj, infj, '[ 0.+infj]')
[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9]
array([0.41337804458542926, 0.9634011951702381 , 0.38924457984371086,
       0.6889281918519422 ])
array([0.41337803, 0.9634012 , 0.3892446 , 0.6889282 ], dtype=float32)
array([0.4133, 0.9634, 0.3892, 0.689 ], dtype=float16)
array([1.e+00, 1.e+01, 1.e+02, 1.e+03, 1.e+04, 1.e+05, 1.e+06, 1.e+07,
       1.e+08, 1.e+09, 1.e+10, 1.e+11, 1.e+12, 1.e+13, 1.e+14, 1.e+15,
       1.e+16, 1.e+17, 1.e+18, 1.e+19])
```

Output with master:
```
(infj, inf*j, '[ 0.+infj]')
[ 0.          0.1         0.2         0.30000001  0.40000001  0.5
  0.60000002  0.69999999  0.80000001  0.89999998]
array([ 0.5864939 ,  0.15714977,  0.26403041,  0.7634838 ])
array([ 0.58649391,  0.15714976,  0.2640304 ,  0.76348382], dtype=float32)
array([ 0.58642578,  0.15710449,  0.26391602,  0.76367188], dtype=float16)
array([  1.00000000e+00,   1.00000000e+01,   1.00000000e+02,
         1.00000000e+03,   1.00000000e+04,   1.00000000e+05,
         1.00000000e+06,   1.00000000e+07,   1.00000000e+08,
         1.00000000e+09,   1.00000000e+10,   1.00000000e+11,
         1.00000000e+12,   1.00000000e+13,   1.00000000e+14,
         1.00000000e+15,   1.00000000e+16,   1.00000000e+17,
         1.00000000e+18,   1.00000000e+19])
```

eric-wieser commented:
Out of curiosity - what keeps us from directly using the C++ version?

eric-wieser commented:
```
array([0.41337803, 0.9634012 , 0.3892446 , 0.6889282 ], dtype=float32)
```
looks kinda bad to me, but I understand the rationale.

What does this give?
```
a = 10.**np.arange(20)
a[0] = 1234
```

ahaldane commented:
Re: the spacing in the float32 array: That is behavior is still to be decided, now that we have more choices due to the existence of a dragon4 interface. Note that example only gets output for `precision='unique'`, but we don't need to make that the default. (In fact, it is not the default in this PR... note I manually set it in my example).

Also, here is what this PR currently outputs by default for your code:
```python
[1]: a = 10.**np.arange(20)
[2]: a[0] = 123456
[3]: a
array([1.23456e+05, 1.00000e+01, 1.00000e+02, 1.00000e+03, 1.00000e+04,
       1.00000e+05, 1.00000e+06, 1.00000e+07, 1.00000e+08, 1.00000e+09,
       1.00000e+10, 1.00000e+11, 1.00000e+12, 1.00000e+13, 1.00000e+14,
       1.00000e+15, 1.00000e+16, 1.00000e+17, 1.00000e+18, 1.00000e+19])
```

ahaldane commented:
Also, another comment on the spacing in `unique` mode:

We already get funny spacing like that in current numpy master:
```python
[31]: array([0.32254803, 0.8584671, 0.07836795, 0.2622], dtype=float32)
array([ 0.32254803,  0.8584671 ,  0.07836795,  0.2622    ], dtype=float32)
```

ahaldane commented:
As for leaving it in C++, I honestly didn't consider that carefully since I assumed it would be harder/more confusing that porting to C and trying to interface to it.  Also, while I didn't modify the dragon4 code itself much, I did pretty heavily rewrite the "wrapper" part of the code that converts dragon4 output to a string.

ahaldane commented:
More thoughts on spacing: I don't think padding with 0s is correct, since the "unnecessary" digits which were dropped are not necessarily 0.

To avoid the trailing spaces, probably we could do as follows: In the first pass in FloatingFormat, figure out the maximum fractional length of the data output will all the trimming. Then, set the precision to that value in the second pass, and don't do any trimming in the second pass (trim-mode 'k'), thus keeping the extra digits in the values that were a bit shorter in the first pass.

mdickinson commented:
Presumably this would fix #9360? (And also close #2643 and #6136.)

charris commented:
The test failures are due to ieee extended precision on 32 linux being aligned on 4 byte boundaries, hence `float96` instead of `float128`.

ahaldane commented:
Yes, I thought I accounted for that with the `#ifdef NPY_FLOAT96`, but apparently that macro-constant isn't defined...

ahaldane commented:
Also I figured out the `1e23` strangeness: It is because the implementation currently doesn't use IEEE unbiased rounding. It should be easy to modify it to do so.

Also, in principle it is possible to change the rounding mode, see [gcc's discussion (link)[https://www.gnu.org/software/libc/manual/html_node/Rounding.html]. However, I think CPython has forced/assumed the rounding mode to be IEEE unbiased, so I think it is safe to assume the same for numpy.

charris commented on numpy/core/src/multiarray/dragon4.h:
```diff
@@ -0,0 +1,68 @@
+
+typedef enum TrimMode
+{
+    TrimMode_None,         // don't trim zeros, always leave a decimal point
```
Should convert these to C style comments.

charris commented:
> Yes, I thought I accounted for that with the #ifdef NPY_FLOAT96, but apparently that macro-constant isn't defined...

Missing include statement. Need `#include <numpy/npy_common.h>`, maybe something in `setup` also, depending

mdickinson commented:
> CPython has forced/assumed the rounding mode to be IEEE unbiased

Assumed, rather than forced. The dtoa.c code is a bit of a mix: some parts (mainly the integer arithmetic parts) explicitly construct the round-ties-to-even results, without help from the FPU. Others (the fast paths that use floating-point), assume that the FPU control word remains at its default of round-ties-to-even. If the FPU rounding mode were set to something else, it's probable that Python's `float` repr would give wrong results in some corner cases, and conceivable that it would do something even more horrible, like entering an infinite loop. On the one hand, that's probably bad, and CPython should (but doesn't) make an effort at least to check that the FPU control word is doing the right thing. On the other hand, it doesn't _seem_ to have caused any error reports to date, and the code has been in use for a good number of years now.

So yes, while in principle assuming that the x87/SSE2 rounding mode settings will be the right ones makes me a bit nervous, in practice it appears to be a fairly safe assumption.

Also, the dtoa.c code is written in such a way that infinite loops could happen (at one point I started refactoring to eliminate the potential for infinite loops, but that project never got completed); I'd hope that the Dragon4 code is a bit better behaved in that respect.

[Full disclosure: I was responsible, with Eric Smith, for implementing the CPython shortest repr implementation, though of course the original core `dtoa.c` code comes from David Gay.]

ahaldane commented:
@mdickinson thanks for the links to the issues.

Assuming IEEE unbiased rounding in this PR is "safe" in the sense that it won't crash. I think the worst case scenario is that we spit out floating point strings that don't covert back to their original values, eg in the case of 1e23. (By the way, I am discussing the rounding mode issues with Ryan Juckett over on his website, linked above).

I don't really want to go through the effort of accounting for all the rounding modes here. But at least maybe I could add a check to `fegetround` during numpy startup that gives a serious warning if the rounding mode isn't unbiased.

ahaldane committed:
ENH: implement IEEE unbiased rounding in dragon4

ahaldane commented:
Updated
 * Implemented IEEE unbiased rounding
 * fixed a possible bug in the Dragon4 implementation
 * did a pass at reformatting everything to match numpy style (including C comments)

When I come back to this I'm going to work on an arrayprint.py interface and on choosing the right defaults. I'm going to try adding a`floatmode` printing option, which controls controls the interpretation of the `precision` printing option. The doc would be:
```
    floatmode : str, optional
        Controls the interpretation of the `precision` option for
        floating-point types. Can take the following values:
            - 'fixed' : Always print exactly `precision` fractional digits,
                    even if this would print more or fewer digits than
                    necessary to specify the value uniquely.
            - 'unique : Print the minimum number of fractional digits necessary
                    to represent each value uniquely. Different elements may
                    have a different number of digits.  The value of the
                    `precision` option is ignored.
            - 'maxprec' : Print at most `precision` fractional digits, but if
                    an element can be uniquely represented with fewer digits
                    only print it with that many.
            - 'maxprec_equal' : Print at most `precision` fractional digits,
                    but if every element in the array can be uniquely
                    represented with an equal number of fewer digits (or less),
                    use that many digits for all elements.
        Default is 'maxprec'
```

ahaldane committed:
ENH: Implement float16 and float128 dragon4 output

ahaldane committed:
ENH: Add Dragon4 printing modes for "unique with cutoff"

ahaldane committed:
ENH: make dragon4 float format functions publicly visible

ahaldane committed:
DOC: update documentation for dragon4

ahaldane committed:
TST: New tests for scalar/array reprs with dragon4

Fixes #9360 #2643 #6136 #9699 #6908
Closes #9919 #9932

ahaldane commented:
Whew. I just spent 2 hours to determine that the pure-Python computation `1e23 == 10.**23` is architecture-dependent. It's `True` on most of our test systems, but `False` on the 2nd, 3rd and 4th appveyor (windows) systems. Probably it's due to some detail of the OS `powf` implementation.

So if you want to generate all allowed `float64` powers of 10 (or the closest IEEE-unbiased rounded numbers), doing `10.**arange(308)` doesn't work in windows. It is safer to do `np.array(['1e{}'.format(i) for i in range(308)], dtype='f8')`. On linux the two are equivalent.

Also, windows does not have `fenv.h` so we can't easily check the rounding mode there, so to be pragmatic I will always just assume it is `FT_TONEAREST` without any tests.

ahaldane commented:
All right, I think this is finished. 

This PR now fixes a number of very old float-representation problems, and also fixes up the LongFloat output as promised in time for 1.14, by generalizing the printing code for any float size. You can get an overview of the PR from the release notes, from the docstrings of `float_format_*`, and from the new tests (2nd to last commit).

After chatting with Ryan Juckett on his website I'm convinced that the core dragon4 implementation is correct, including the few small modifications that Ryan made relative to the Burger & Dybvig paper, plus some small tweaks of my own for IEEE unbiased rounding.

I've tried to arrange the commits in a way that shows which parts were just the port from C++ to C, and which parts were modifications. It might be easier to understand what was changed by going through the commits one by one instead of looking at all the changes at once. Let me know if I can reorganize it in some way that is helpful.

charris commented:
> Probably it's due to some detail of the OS powf implementation.

Kudos tracking that down. Sometimes I think NumPy is a better test of math libraries than the actual library maintainers :)

charris commented:
I haven't reviewed in detail, but this looks pretty clean and I suspect you have a far better understanding of its contents than I, so let's get it in. Thanks Allan.

ahaldane commented:
Whew that was quick!

I'll go through and close the relevant PRs, I think I needed commas in the commit message because they were not auto-closed.

charris commented:
I think you need a separate `closes` for each issue.

charris commented:
The rounding still worries me and might be a problem. I suppose we need to rely on the tests to check if that is a problem. Also, out of curiosity, what were the python (hence compiler) versions that had trouble with 1e23?

eric-wieser commented on numpy/core/tests/test_scalarprint.py:
```diff
@@ -5,26 +5,193 @@
+        assert_equal(fpos64('3.14159265358979323846'), "3.141592653589793")
+
+
+	# smallest numbers
```
nit: bad indentation here

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -463,6 +483,23 @@ def array2string(a, max_line_width=None, precision=None,
         (whitespace character) in the sign position of positive values.  If
         '-', omit the sign character of positive values. If 'legacy', print a
         space for positive values except in 0d arrays.
+    floatmode : str, optional
```
Are we sure we don't want `float_mode` for consistency with `suppress_small`?

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -736,75 +937,22 @@ def __call__(self, x):
-            return format_longfloat(x, self.precision)
-
-
-class LongComplexFormat(object):
```
I think we might need to keep this for compatibility, as a subclass that does nothing but throw a warning in it's constructor.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -597,8 +634,8 @@ def _formatArray(a, format_function, rank, max_line_len,
 
-class FloatFormat(object):
-    def __init__(self, data, precision, suppress_small, sign=False):
+class FloatingFormat(object):
```
Same comment about leaving a public `FloatFormat` object visible, assuming we're considering this public API.

Docstring saying `For np.floating objects` or something

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -736,75 +937,22 @@ def __call__(self, x):
-    # XXX Have to add something to determine the width to use a la FloatFormat
-    # Right now, things won't line up properly
-    def __init__(self, precision, sign=False):
+class ComplexFormat(object):
```
Maybe `ComplexFloatingFormat` to match `np.complexfloating`

eric-wieser reviewed (commented):
Really pleased to see all the arrayprint formatters merge together, but a couple nits about names and compatibility in `arrayprint`, which we should probably look at before 1.14

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -828,107 +712,107 @@ timedeltatype_str(PyObject *self)
 
-#define PREC @NAME@PREC_@KIND@
+static PyObject *
+@name@type_@kind@_either(npy_@name@ val, TrimMode trim_pos, TrimMode trim_sci,
```
Could do with a comment explaining what this function is for

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -828,107 +712,107 @@ timedeltatype_str(PyObject *self)
-static int
-c@name@type_print(PyObject *v, FILE *fp, int flags)
+static PyObject *
+halftype_@kind@(PyObject *self)
```
Is this special compared to the other ones? Why?

eric-wieser commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -290,29 +290,110 @@ def test_sign_spacing(self):
+        assert_equal(repr(y),
+    "array([0.2918820979355541 , 0.5064172631089138 , 0.2848750619642916 ,\n"
+    "       0.4342965294660567 , 0.7326538397312751 , 0.3459503329096204 ,\n"
+    "       0.0862072768214508 , 0.39112753029631175])")
```
Indentation is pretty funky here too

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -321,15 +347,9 @@ def _get_format_function(data, **options):
+        return formatdict['float']()
     elif issubclass(dtypeobj, _nt.complexfloating):
-        if issubclass(dtypeobj, _nt.clongfloat):
-            return formatdict['longcomplexfloat']()
```
This means that users setting a `longcomplexfloat` formatter will now see it being ignored silently

eric-wieser commented:
Deleted my comments about whitespace, and just added a STY commit to #9962, to reduce noise here
